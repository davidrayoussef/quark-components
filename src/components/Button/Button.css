.button {
  position: relative;
  padding: 0 15px;
  text-align: center;
  color: #faf9f9;
  cursor: pointer;
  border: 0;
  transition: 0.2s;
  overflow: hidden;
}

.primary {
  background: rgba(3, 169, 244, 1);
}

.primary:hover {
  background: rgba(3, 169, 244, 0.8);
}

.secondary {
  background: rgba(0, 0, 0, 0.3);
}

.secondary:hover {
  background: rgba(0, 0, 0, 0.4);
}

.reverse {
  background: rgba(0, 0, 0, 0.02);
  color: #666;
  box-shadow: inset 0 0 0 1.5px rgb(150, 203, 254);
}

.reverse:hover {
  background: rgba(0, 0, 0, 0.05);
}

.danger {
  background: rgba(244, 74, 56, 1);
}

.danger:hover {
  background: rgba(244, 74, 56, 0.8);
}

.small {
  font-size: 12px;
  padding: 8px;
}

.normal {
  font-size: 15px;
  padding: 10px 15px;
}

.large {
  font-size: 16px;
  padding: 15px 20px;
}

.button:disabled,
.button[disabled] {
  background: #cccccc;
  pointer-events: none;
}

/* For ripple effect */
.button:after {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  width: 10px;
  height: 10px;
  background: rgba(255, 255, 255, 0.15);
  opacity: 0;
  border-radius: 100%;
  transform: scale(1, 1) translate(-50%);
  transform-origin: 50% 50%;
}

.button.reverse:after {
  background: rgba(0, 0, 0, 0.1);
}

@keyframes ripple {
  0% {
    transform: scale(0, 0);
    opacity: 1;
  }
  20% {
    transform: scale(12, 12);
    opacity: 1;
  }
  100% {
    opacity: 0;
    transform: scale(25, 25);
  }
}

.button:focus:not(:active)::after {
  animation: ripple 1s ease-out;
}
