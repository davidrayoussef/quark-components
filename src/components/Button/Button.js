import React from 'react';
import PropTypes from 'prop-types';
import style from './Button.css';

const Button = ({ children, type, size, ...props }) => {
  const getTypeStyle = (type, style) => `
    ${
      type === 'secondary' ? style.secondary :
      type === 'reverse' ? style.reverse :
      type === 'danger' ? style.danger :
      style.primary
    }
  `;

  return (
    <button
      className={`
        ${style.button}
        ${getTypeStyle(type, style)}
      `}
      {...props}
    >
      { children }
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.node,
  type: PropTypes.string,
  size: PropTypes.string
};

export default Button;
