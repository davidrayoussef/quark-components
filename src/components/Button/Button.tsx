import React from 'react';

import { ButtonProps } from '@/shared';
import { getTypeStyle, getSizeStyle } from '@/utilities';

import style from './Button.scss';

export const Button: React.FC<ButtonProps> = ({
  children,
  styleType,
  size,
  component,
  ...rest
}: ButtonProps) => {
  const Component = component ?? 'button';
  return (
    <Component
      className={`
        ${style.button}
        ${getTypeStyle(styleType, style)}
        ${getSizeStyle(size, style)}
      `}
      {...rest}
    >
      {children}
    </Component>
  );
};
