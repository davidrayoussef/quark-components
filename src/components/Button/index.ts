export { Button } from './Button';
export { RouterLinkButton } from './RouterLinkButton';
