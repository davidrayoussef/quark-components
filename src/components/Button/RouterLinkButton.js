import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import Button from './Button';

const RouterLinkButton = ({ history, location, match, staticContext, to, onClick, /* <-- remove unusable props */ ...rest }) => {
  return (
    <Button
      {...rest}
      size="small"
      onClick={(e) => {
        history.push(to);
      }}
    />
  );
};

RouterLinkButton.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  staticContext: PropTypes.string,
  to: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  children: PropTypes.node
};

export default withRouter(RouterLinkButton);
