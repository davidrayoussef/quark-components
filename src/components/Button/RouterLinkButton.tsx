import React from 'react';
import { withRouter } from 'react-router';

import { Button } from './Button';

import { ButtonSize, LinkButtonProps } from '../../shared';

const LinkButton: React.FC<LinkButtonProps> = ({
  history,
  location,
  match,
  staticContext,
  to,
  children,
  onClick,
  ...rest
}: LinkButtonProps) => {
  return (
    <Button
      {...rest}
      size={ButtonSize.Small}
      onClick={() => {
        history.push(to);
      }}
    >
      {children}
    </Button>
  );
};

export const RouterLinkButton = withRouter(LinkButton);
