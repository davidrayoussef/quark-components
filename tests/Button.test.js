import React from 'react';
import { shallow } from 'enzyme';
import Button from '../src/components/Button';

test('should render', () => {
  const wrapper = shallow(
    <Button></Button>
  );
  expect(wrapper);
});

test('should render correct Button text', () => {
  const wrapper = shallow(
    <Button>Button Text</Button>
  );
  expect(wrapper.text()).toBe('Button Text');
});

test('should add .primary class if no type is passed', () => {
  const wrapper = shallow(
    <Button>Button Text</Button>
  );
  expect(wrapper.find('.primary').length).toBe(1);
});

test('should add .primary class if styleType="primary"', () => {
  const wrapper = shallow(
    <Button styleType="primary">Button Text</Button>
  );
  expect(wrapper.find('.primary').length).toBe(1);
});

test('should add .secondary class if styleType="secondary"', () => {
  const wrapper = shallow(
    <Button styleType="secondary">Button Text</Button>
  );
  expect(wrapper.find('.secondary').length).toBe(1);
});

test('should add .reverse class if styleType="reverse"', () => {
  const wrapper = shallow(
    <Button styleType="reverse">Button Text</Button>
  );
  expect(wrapper.find('.reverse').length).toBe(1);
});

test('should add .danger class if styleType="danger"', () => {
  const wrapper = shallow(
    <Button styleType="danger">Button Text</Button>
  );
  expect(wrapper.find('.danger').length).toBe(1);
});

test('snapshot', () => {
  const wrapper = shallow(
    <Button styleType="primary">Button Text</Button>
  );
  expect(wrapper).toMatchSnapshot();
});
