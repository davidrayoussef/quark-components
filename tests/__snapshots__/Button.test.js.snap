// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<button
  className="
        button
        
  primary

        
  normal

      "
>
  Button Text
</button>
`;
